### PR TITLE
zip*/unzip methods added #487

### PR DESCRIPTION
**Pull request by ruslansennov**

### CODE REVIEW

File: src/main/java/javaslang/collection/Iterator.java [hunk 1]
```diff
@@ -323,7 +324,36 @@ public T next() {
 
     @Override
     default Iterator<T> dropWhile(Predicate<? super T> predicate) {
-        return null;
+        if (!hasNext()) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Iterator.java]: let's do `Objects.requireNonNull(predicate, "predicate is null")`. I've also forgotten in in various places and will add them later...

File: src/main/java/javaslang/collection/Iterator.java [hunk 2]
```diff
@@ -694,6 +724,83 @@ public T next() {
         return null;
     }
 
+    default <U> Iterator<Tuple2<T, U>> zip(Iterable<U> that) {
+        if(isEmpty()) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Iterator.java]: dito

File: src/main/java/javaslang/collection/Iterator.java [hunk 3]
```diff
@@ -694,6 +724,83 @@ public T next() {
...
+                    if (!hasNext()) {
+                        EMPTY.next();
+                    }
+                    return Tuple.of(it1.next(), it2.next());
+                }
+            };
+        }
+    }
+
+    default <U> Iterator<Tuple2<T, U>> zipAll(Iterable<U> that, T thisElem, U thatElem) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Iterator.java]: dito

File: src/main/java/javaslang/collection/Iterator.java [hunk 4]
```diff
@@ -694,6 +724,83 @@ public T next() {
         return null;
     }
 
+    default <U> Iterator<Tuple2<T, U>> zip(Iterable<U> that) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Iterator.java]: dito

File: src/main/java/javaslang/collection/HashSet.java [hunk 5]
```diff
@@ -181,15 +181,15 @@ public boolean contains(T element) {
     @Override
     public HashSet<T> dropWhile(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
-        final List<T> dropped = list.get().dropWhile(predicate);
-        return dropped.length() == list.get().length() ? this : HashSet.ofAll(dropped);
+        final HashSet<T> dropped = HashSet.ofAll(iterator().dropWhile(predicate));
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashSet.java]: Yay! Getting rid of list :-)

Comment (ruslansennov) [inline src/main/java/javaslang/collection/HashSet.java]: :smile:

**Comment by danieldietrich** [comment]
thx!

**Pull request merged**